src/libmeasurement_kit/ooni/captive_portal.cpp

```diff
@@ -0,0 +1,231 @@
+                             unfiltered ? "yes" : "no");
+                result["result"] = unfiltered;
+            }
+            (*entry)["vendor_http_tests"].push_back(result);
+            done_cb(err);
+        };
+    };
+
+    std::vector<Continuation<Error>> continuations;
+    // Note: regarding `input` we're making copies on purpose for
+    // robustness. We can be more smart using move semantic, but it
+    // doesn't seem we need to optimize that much (it's small).
+    for (auto input : gen_http_inputs()) {
+        logger->info("setting up %s", input.at("name").c_str());
+        options["http/url"] = input.at("url");
+        std::string body;
+        http::Headers headers;
+        if (input.count("ua")) {
+            headers = {{"User-Agent", input.at("ua")}};
+        } else {
+            headers = constants::COMMON_CLIENT_HEADERS;
+        }
+
+        // Note: the following callback uses `[=]`, which means that the
+        // options are copied in each run. Thus, each test is getting its
+        // right URL to connect to and check.
+        continuations.push_back([=](Callback<Error> done_cb) {
+            templates::http_request(entry, options, headers, body,
+                                    http_cb(input, done_cb), reactor, logger);
+        });
+    }
+    mk::parallel(continuations, all_done_cb, 3);
+}
+
+// this is the only test that doesn't follow the pattern of those above.
+// this hostname should always resolve to this IP.
+void dns_msft_ncsi(Var<Entry> entry, Callback<Error> done_cb, Settings options,
+                   Var<Reactor> reactor, Var<Logger> logger) {
+    std::string hostname = "dns.msftncsi.com";
+    // Note: we're setting the nameserver to empty, which is going to work
+    // as long as we're using the `system` DNS resolver
+    std::string nameserver = "";
+    templates::dns_query(entry, "A", "IN", hostname, nameserver,
+                         [=](Error err, Var<dns::Message> message) {
+                             Entry result = {{"name", "Microsoft DNS"},
+                                             {"result", nullptr},
+                                             {"hostname", "dns.msftncsi.com"},
+                                             {"expected_ip", "131.107.255.255"},
+                                             {"failure", nullptr}};
+                             if (!!err) {
+                                 logger->info("dns_query err: %s",
+                                              err.as_ooni_error().c_str());
+                                 result["failure"] = err.as_ooni_error();
+                                 done_cb(err);
+                             } else {
+                                 for (const auto &a : message->answers) {
+                                     // XXX: This should actually be adapted
+                                     // to also deal with IPv6 replies
```

Comment:
As the code is currently written, each non-IPv4 entry will appear as an empty string inside of the `actual_ips` list.